### ISSUE TITLE
Error: TCP Timeout

### ISSUE THREAD

**Issue by davilico05** [plasticrake/homebridge-tplink-smarthome#44]
### errors when starting hombridge

*** WARNING *** The program 'node' uses the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node>
*** WARNING *** The program 'node' called 'DNSServiceRegister()' which is not supported (or only supported partially) in the Apple Bonjour compatibility layer of Avahi.
*** WARNING *** Please fix your application to use the native API of Avahi!
*** WARNING *** For more information see <http://0pointer.de/avahi-compat?s=libdns_sd&e=node&f=DNSServiceRegister>

### Versions

- **Node**:       v8.9.1
- **Homebridge**: v0.4.33

### Configuration (config.json)
```json
{
	"bridge": {
		"name": "Homebridge",
		"username": "CC:22:3D:E3:CE:30",
		"port": 51826,
		"pin": "031-45-154"
	},

	"platforms": [{
		"platform": "TplinkSmarthome",
		"name": "TplinkSmarthome",
		"broadcast": "192.168.0.24",
		"devices": ["HS100"],
		"deviceTypes": ["plug"],
		"macAddresses": ["50:c7:bf:ec:f0:f5"],
		"pollingInterval": 10,
		"addCustomCharacteristics": false,
		"inUseThreshold": 0,
		"switchModels": ["HS100"],
		"timeout": 5
	}],

	"accessories": []
}
```
### PROBLEM ###

I can not solve this connection problem.

The HS100 is configured with the Kasa app and with the app it works without problems. That is, it is VERY SLOW.

In the console of the RP3, I constantly get the timeout message.

When access to Home app and it appears "Updating" get this message on the console RP3 and the Home App appears "No Answer". Repeats in Loop.

### LOG ###
```
[2017-11-24 21:48:49] [TplinkSmarthome.API] TCP Timeout
[2017-11-24 21:48:49] [TplinkSmarthome.API] [Flexo] device.send() Error: TCP Timeout
[2017-11-24 21:48:49] [TplinkSmarthome] [Flexo] getPowerState
[2017-11-24 21:48:49] [TplinkSmarthome] Error: TCP Timeout
    at Timeout._onTimeout (/usr/lib/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js:249:20)
    at ontimeout (timers.js:475:11)
    at tryOnTimeout (timers.js:310:5)
    at Timer.listOnTimeout (timers.js:270:5)
```

**Comment by plasticrake** [comment]
I see you are setting a lot of config options instead of using the default. Why is that? Your broadcast address looks wrong as well. You are setting the timeout to 5 seconds, but that is the default. Did you try setting it higher to 10 or 20? 

try something simple like:
```json
	"platforms": [{
		"platform": "TplinkSmarthome",
		"name": "TplinkSmarthome",
		"timeout": 20
	}],
```

If this does not help please include more from your log.

**Comment by davilico05** [comment]
In another attempt I put "timeout": 9000

I will carry out tests and send you the log again. Even without configuring the config.json the timeout error persists.

I even tried to factory reset the tp-link plug

**Comment by sean-co** [comment]
I seem to have a similar issue. Usually happens when trying to set a scene with a few light bulbs at one time. If you tap on the scene to run it again then it usually goes through with no more errors.

This is the config.json setting i'm using currently. I have tried without the "timout 20" setting and also without the "devices" defined as they are in this picture. (I know the subnets are different between the logs and this picture, i'm capturing these logs while at work, same problem persists while on the same subnet at home).
![image](https://user-images.githubusercontent.com/34090414/33373135-35e965ea-d509-11e7-8278-6d0ede61b582.png)

The IP addresses listed in the above picture are all reserved on my router
![image](https://user-images.githubusercontent.com/34090414/33373460-a700297a-d50a-11e7-814f-59a114021d8b.png)

This is the log as i got the timeout

Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '12.11' with data: {"characteristics":[{"aid":12,"iid":11,"value":true}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Muting event '17.12' notification for this connection since it originated here.
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Muting event '14.12' notification for this connection since it originated here.
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '15.11' with data: {"characteristics":[{"aid":15,"iid":11,"value":true}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Muting event '13.12' notification for this connection since it originated here.
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '16.11' with data: {"characteristics":[{"aid":16,"iid":11,"value":true}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '14.11' with data: {"characteristics":[{"aid":14,"iid":11,"value":true}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '14.12' with data: {"characteristics":[{"aid":14,"iid":12,"value":1}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '14.13' with data: {"characteristics":[{"aid":14,"iid":13,"value":366}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '17.11' with data: {"characteristics":[{"aid":17,"iid":11,"value":true}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '17.12' with data: {"characteristics":[{"aid":17,"iid":12,"value":1}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Muting event '13.13' notification for this connection since it originated here.
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '18.11' with data: {"characteristics":[{"aid":18,"iid":11,"value":true}]}
Wed, 29 Nov 2017 11:20:47 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '18.13' with data: {"characteristics":[{"aid":18,"iid":13,"value":366}]}
[11/29/2017, 1:20:52 PM] [TplinkSmarthome.API] UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome.API] [Kitchen 1] device.send() Error: UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome] [Kitchen 1] setLightState
[11/29/2017, 1:20:52 PM] [TplinkSmarthome] Error: UDP Timeout
    at Timeout._onTimeout (/usr/local/lib/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js:170:30)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5)
Wed, 29 Nov 2017 11:20:52 GMT Accessory [Homebridge] Error setting Characteristic "On" to value 1:  UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome.API] UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome.API] UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome.API] UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome.API] [Kitchen 2] device.send() Error: UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome.API] [Bedroom] device.send() Error: UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome.API] [Front Door] device.send() Error: UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome] [Kitchen 2] setLightState
[11/29/2017, 1:20:52 PM] [TplinkSmarthome] Error: UDP Timeout
    at Timeout._onTimeout (/usr/local/lib/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js:170:30)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5)
Wed, 29 Nov 2017 11:20:52 GMT Accessory [Homebridge] Error setting Characteristic "Color Temperature" to value 366:  UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome] [Bedroom] setLightState
[11/29/2017, 1:20:52 PM] [TplinkSmarthome] Error: UDP Timeout
    at Timeout._onTimeout (/usr/local/lib/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js:170:30)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5)
Wed, 29 Nov 2017 11:20:52 GMT Accessory [Homebridge] Error setting Characteristic "Color Temperature" to value 366:  UDP Timeout
[11/29/2017, 1:20:52 PM] [TplinkSmarthome] [Front Door] setLightState
[11/29/2017, 1:20:52 PM] [TplinkSmarthome] Error: UDP Timeout
    at Timeout._onTimeout (/usr/local/lib/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js:170:30)
    at ontimeout (timers.js:386:14)
    at tryOnTimeout (timers.js:250:5)
    at Timer.listOnTimeout (timers.js:214:5)
Wed, 29 Nov 2017 11:20:52 GMT Accessory [Homebridge] Error setting Characteristic "Brightness" to value 50:  UDP Timeout
Wed, 29 Nov 2017 11:20:52 GMT EventedHTTPServer [::ffff:172.18.202.86] HTTP Response is finished
Wed, 29 Nov 2017 11:20:52 GMT EventedHTTPServer [::ffff:172.18.202.86] Writing pending HTTP event data
Wed, 29 Nov 2017 11:20:55 GMT EventedHTTPServer [::ffff:172.18.202.86] Sending HTTP event '13.13' with data: {"characteristics":[{"aid":13,"iid":13,"value":154}]}

Setting a deice one at a time seems to work just fine.
![image](https://user-images.githubusercontent.com/34090414/33373693-bf21fbe0-d50b-11e7-8ae5-95d3e0989a7a.png)

**Comment by NajibNour** [comment]
Was this problem solved?

**Comment by DBCraft01** [comment]
I cloud solve it just by reseting my Fritz-Wlan-Repeater and my Fritz-Box-Router and it´s working just fine again.

**Comment by UNiXMIT** [comment]
Not sure if this thread was solved but I am getting a similar message. I'm running the latest version 3.11.0
Runs ok for a short while then I get this so I have to restart hombridge.
`Jun 04 15:30:31 unix homebridge[2501]: [6/4/2018, 3:30:31 PM] [TplinkSmarthome] [Lounge1] getLightState`
`Jun 04 15:30:31 unix homebridge[2501]: [6/4/2018, 3:30:31 PM] [TplinkSmarthome] Error: TCP Timeout`
`Jun 04 15:30:31 unix homebridge[2501]: at Timeout._onTimeout (/usr/local/lib/node_modules/homebridge-tplink-smarthome/node_modules/tplink-smarthome-api/lib/client.js:249:20)`
`Jun 04 15:30:31 unix homebridge[2501]: at ontimeout (timers.js:386:14)`
`Jun 04 15:30:31 unix homebridge[2501]: at tryOnTimeout (timers.js:250:5)`
`Jun 04 15:30:31 unix homebridge[2501]: at Timer.listOnTimeout (timers.js:214:5)`
`Jun 04 15:30:31 unix homebridge[2501]: [6/4/2018, 3:30:31 PM] [TplinkSmarthome.API] TCP Timeout`
`Jun 04 15:30:31 unix homebridge[2501]: [6/4/2018, 3:30:31 PM] [TplinkSmarthome.API] [Lounge1] device.send() Error: TCP Timeout`
`Jun 04 15:30:31 unix homebridge[2501]: (node:2501) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 1): Error: TCP Timeout`
`Jun 04 15:30:31 unix homebridge[2501]: (node:2501) DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.`

**Comment by UNiXMIT** [comment]
Ok that didn't work. Trying "timeout": 300

**Comment by UNiXMIT** [comment]
I guess i'll have to open a new thread for my issue. This one is dead.

**Comment by lay55** [comment]
I've been having the same issue


This is my log:

[2018-11-28 19:10:15] [TplinkSmarthome.API] [My Smart Switch] device.send() Error: connect EHOSTUNREACH 192.168.1.73:9999 - Local (192.168.1.5:51685)
[2018-11-28 19:10:15] [TplinkSmarthome] [My Smart Switch] setPowerState
[2018-11-28 19:10:15] [TplinkSmarthome] { Error: connect EHOSTUNREACH 192.168.1.73:9999 - Local (192.168.1.5:51685)
    at internalConnect (net.js:964:16)
    at defaultTriggerAsyncIdScope (internal/async_hooks.js:281:19)
    at net.js:1062:9
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickCallback (internal/process/next_tick.js:181:9)
  errno: 'EHOSTUNREACH',
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.1.73',
  port: 9999 }
[2018-11-28 19:10:16] [TplinkSmarthome.API] [My Smart Switch] device.send() Error: connect EHOSTDOWN 192.168.1.73:9999 - Local (192.168.1.5:51686)
[2018-11-28 19:10:16] [TplinkSmarthome] [My Smart Switch] setPowerState
[2018-11-28 19:10:16] [TplinkSmarthome] { Error: connect EHOSTDOWN 192.168.1.73:9999 - Local (192.168.1.5:51686)
    at internalConnect (net.js:964:16)
    at defaultTriggerAsyncIdScope (internal/async_hooks.js:281:19)
    at net.js:1062:9
    at _combinedTickCallback (internal/process/next_tick.js:132:7)
    at process._tickCallback (internal/process/next_tick.js:181:9)
  errno: 'EHOSTDOWN',
  code: 'EHOSTDOWN',
  syscall: 'connect',
  address: '192.168.1.73',
  port: 9999 }


And this if my config.json:


{
            "platform": "TplinkSmarthome",
            "name": "TplinkSmarthome",
            "devices": [
                {
                    "host": "192.168.1.73"
                }
            ]
        }

**Comment by xrv1ck** [comment]
Regarding OP's log file, I was running into the exact same issue on a pi4 using Raspbian Buster. I running the latest version of homebridge (4.5) and homebridge-tplink-smarthome (4.0.1).

The bug was driving me nuts as I just upgraded from a Pi 1 and everything worked fine before that. I finally decided to downgrade my homebridge-tplink-smarthome plugin down to 3.13.0 just to see if it would fix it and VIOLA. I didn't explore much further into which version this bug was specifically introduced but at least for now, I'm up and running with no problems.

TLDR: Downgrade to 3.13.0 and you should be good to go.

**Comment by Staubgeborener** [comment]
@xrv1ck @plasticrake 
> TLDR: Downgrade to 3.13.0 and you should be good to go.
I can not confirm this. I use also raspbian buster on a pi 4 and downgraded to version 3.13.0, still same issue with the timeout. 

Interesting fact: I can control my tplink HS110 with Eve and Homekit (turn on and off), but no chance to get the current values (and like i said, still timeouts).

**Comment by xrv1ck** [comment]
Testing it more after I redid my homebridge set up again, I can also confirm the fix didn’t work. Sorry about that! 

But I did I find out that unplugging and plugging back in my HS300 smart power strip would make the error messages go away. My guess is that there is there’s some network connectivity issues with the particular device producing the error messages and having it connect to homebridge. 

Hopefully this solves it for you as well.

**Comment by Staubgeborener** [comment]
Nope, didn't solves. I already unplugged several times my hs110. Its so weird, i'll get my values (like current) in my eve app, but i can't create an automation with that. I mean - i can, but its like, there is no automation. It will simply be ignored (and its a simple one, like "if light on and current of hs110 over 1W, turn light off"). And again, there is this timeout error.
I simply can't find the problem. The hs110 also has a a static IP.

**Comment by stephenjmcmahon** [comment]
I'm also hitting the HOSTUNREACHABLE issue with the KP400 (which stays connected 24/7) on a static IP. Kind of an odd error... I'm assuming there is no route from the local system, and the packets to be sent were never put on the wire. Not sure why it's available only sometimes

CONFIG:
```
        {
            "platform": "TplinkSmarthome",
            "name": "TplinkSmarthome",
            "switchModels": "KP400",
            "devices": [
                {
                    "host": "192.168.60.26",
                    "port": "9999"
                }
```

ERROR LOG:
```
[1/22/2020, 02:37:48] [TplinkSmarthome] [Patio Lights]
[1/22/2020, 02:37:48] [TplinkSmarthome] Error: connect EHOSTUNREACH 192.168.60.26:9999
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1129:14) {
  errno: 'EHOSTUNREACH',
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.60.26',
  port: 9999
}
[1/22/2020, 02:37:48] [TplinkSmarthome.API] [Unused Patio Outlet] device.send() Error: connect EHOSTUNREACH 192.168.60.26:9999
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1129:14) {
  errno: 'EHOSTUNREACH',
  code: 'EHOSTUNREACH',
  syscall: 'connect',
  address: '192.168.60.26',
  port: 9999
```

**Comment by plasticrake** [comment]
Closing due to inactivity. Please reopen a new issue if there is still an issue on the newest version of the plugin.
